### PR TITLE
More compact representation of `DomainName`

### PR DESCRIPTION
**Pull request by kim**
# Description of Changes

It seemed at some point feasible to pass a custom `Ord` dictionary to STDB _somehow_, but it isn't. As `DomainName` gets lowered to a `String` anyway, we can represent it more compactly as a `String` + offset to where the TLD part ends.

# API

 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

File: crates/lib/src/name.rs [hunk 1]
```diff
@@ -287,69 +266,56 @@ impl From<Tld> for DomainPart {
...
-    /// Drop subdomain, if any, and return only the TLD
-    pub fn into_tld(self) -> Tld {
-        self.tld
+    pub fn tld(&self) -> &TldRef {
+        TldRef::new(&self.domain_name[..self.tld_offset])
     }
 
-    pub fn sub_domain(&self) -> Option<&DomainPart> {
-        self.sub_domain.as_ref()
+    pub fn as_tld(&self) -> Tld {
```
[the remark this answers is not in the document]
Comment (kim) [reply crates/lib/src/name.rs:291]: I beg to differ: analogous to the `Into` trait, `into_*` inherents usually consume the receiver (which is not the case here). The difference between `as_*` and `to_*` is often less clear, but `to_` usually suggests that allocation may happen. Since that's the case here, I'd be fine with `to_tld`.

File: crates/lib/src/name.rs [hunk 2]
```diff
@@ -287,69 +266,56 @@ impl From<Tld> for DomainPart {
...
-    pub fn tld(&self) -> &Tld {
-        &self.tld
+    pub fn as_str(&self) -> &str {
+        &self.domain_name
     }
 
-    /// Drop subdomain, if any, and return only the TLD
-    pub fn into_tld(self) -> Tld {
-        self.tld
+    pub fn tld(&self) -> &TldRef {
```
[the remark this answers is not in the document]
Comment (kim) [reply crates/lib/src/name.rs:287]: Well, morally this is a getter. We don't do `get_` for getters, or do we?

File: crates/lib/src/name.rs [hunk 3]
```diff
@@ -287,69 +266,56 @@ impl From<Tld> for DomainPart {
 /// Each segment in a database name can contain any UTF-8 character, except for
 /// whitespace and '/'. The maximum segment length is 64 characters.
 ///
-/// [`DomainName`]s consist of [`DomainPart`]s, and are compared case-
-/// insensitively. Note, however, that the [`fmt::Display`] and [`Serialize`]
-/// impls will use the original, (potentially) mixed-case representation.
+/// Note that [`PartialEq`] compares the exact string representation of a
+/// [`DomainName`], as one would expect, but the SpacetimeDB registry compares
```
[the remark this answers is not in the document]
Comment (kim) [reply crates/lib/src/name.rs:270]: At some point, there was a -- somewhat handwavy -- desire to "compare database names lowercase", presumably to avoid visually similar names. Since all of UTF-8 (except whitespace) is allowed in a `DomainName`, this is actually non-trivial: do we convert to Unicode normal form (and if so, which)? Do we apply proper case-folding? Or compare character-wise? Or just `to_lowercase`, which respects _some_ rules regarding the positions of characters within a word? Does that catch everything "visually similar", and what is the threat model actually?

On the other hand, it seems quite desirable to preserve the exact casing as given by the user for display purposes. 

Originally, the idea was to make `PartialEq` canonical for both a "plain" value and a SpacetimeDB column value. But alas, this doesn't work because we don't have custom comparators (yet)! 

So we need to store both the original value and a `String` key in exactly the form so `PartialEq` for `String` yields the desired comparison. Before this change, this was exactly what was happening, at least doubling the size of `DomainName` when the lowercase version was not needed most of the time.

**tl;dr** A `DomainName` is just a fancy string most of the time, except for the SpacetimeDB DNS. Because Unicode is hard, we may want to change the semantics of the DNS down the road -- but we definitely **don't** want `PartialEq` and DNS to diverge! 

If you don't mind, I would just remove the commentary here. It is, if you will, an "implementation detail" of the SpacetimeDB DNS.

**Review by Centril — APPROVED** [review]
